File: app/src/main/java/ru/arhlib/app/afisha/AfishaViewModel.java
package ru.arhlib.app.afisha;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ru.arhlib.app.data.Api;
import ru.arhlib.app.data.AppDatabase;
import ru.arhlib.app.data.Webservice;

public class AfishaViewModel extends AndroidViewModel {
    private LiveData<Page> page;
    private PageDao pageDao;
    private Webservice webservice;

    public AfishaViewModel(@NonNull Application application) {
        super(application);
        pageDao = AppDatabase.getInstance(application).pageDao();
        webservice = Api.getInstance();
    }

    public LiveData<Page> getPage(int pageId) {
        if (page == null) {
            page = pageDao.load(pageId);

            webservice.getPage(pageId).enqueue(new Callback<Page>() {
                @Override
                public void onResponse(Call<Page> call, Response<Page> response) {
                    if (response.isSuccessful()) {
                        new Thread(() -> {
                            pageDao.save(response.body());
                        }).start();
                    }
                }

                @Override
                public void onFailure(Call<Page> call, Throwable t) {
                    call.clone().enqueue(this);
                }
            });
        }
        return page;
    }
}

File: app/src/main/java/ru/arhlib/app/afisha/AfishaViewModel.kt
package ru.arhlib.app.afisha

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import ru.arhlib.app.data.Webservice

class AfishaViewModel : ViewModel() {
    private val webservice: Webservice = Webservice.instance

    val afisha = MutableLiveData<LoadResult<Page>>()

    init {
        load()
    }

    fun load() {
        afisha.postValue(LoadResult.Loading)

        viewModelScope.launch {
            try {
                afisha.postValue(LoadResult.Success(getPage()))
            } catch (throwable: Throwable) {
                afisha.postValue(LoadResult.Error)
            }
        }
    }

    private suspend fun getPage(): Page {
        return webservice.getPage(6671)
    }
}
